src/reverse.zig

```diff
@@ -0,0 +1,151 @@
+const std = @import("std");
+const address = @import("./address.zig");
+const dns = @import("./lib.zig");
+const assert = std.debug.assert;
+
+/// ReverseLookup is the primary interface for classless IN-ADDR.ARPA (IPv4) and IP6.ARPA (IPv6) delegations
+pub const ReverseLookup = struct {
+    allocator: std.mem.Allocator,
+    ip_address: []const u8,
+    packet_id: u16, // Arbitrary packet ID
+
+    const Self = @This();
+    const max_ipv6_label_size = 35;
+    const max_ipv4_label_size = 6;
+
+    pub fn init(allocator: std.mem.Allocator, ip_address: []const u8, packet_id: u16) !Self {
+        return Self{
+            .allocator = allocator,
+            .ip_address = ip_address,
+            .packet_id = packet_id,
+        };
+    }
+
+    /// Reverse lookup on a given ipv4 ip address
+    pub fn lookupIpv4(self: Self) ![][]const u8 {
+        var add = try address.IpAddress.init(self.allocator, try address.AddressMeta.Ipv4().fromString(self.ip_address));
+        const arpa_address = try add.reverseIpv4();
+
+        var labels: [max_ipv4_label_size][]const u8 = undefined;
+        const apra_address_dns_name = try dns.Name.fromString(arpa_address, &labels);
+
+        return try self.buildAndSendPacket(apra_address_dns_name);
+    }
+
+    /// Reverse lookup on a given ipv6 ip address
+    pub fn lookupIpv6(self: Self) ![][]const u8 {
+        var add = try address.IpAddress.init(self.allocator, try address.AddressMeta.Ipv6().fromString(self.ip_address));
+        const arpa_address = try add.reverseIpv6();
+
+        var labels: [max_ipv6_label_size][]const u8 = undefined;
+        const apra_address_dns_name = try dns.Name.fromString(arpa_address, &labels);
+
+        return self.buildAndSendPacket(apra_address_dns_name) catch &[_][]const u8{};
```

Comment:
this should return the error instead of an empty stack-allocated array. currently this makes it not clear who owns the memory given by this function, as an error in `buildAndSendPacket()` will cause a library user to attempt to call `allocator.free()` (given they had to give an allocator to this function) a memory region that is on the stack, causing segmentation fault on non-zero arrays (which this could be, given `[_]`).

example:
```zig
const std = @import("std");

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();
    const coolData = &[1][]const u8{"test"};
    allocator.free(coolData);
}
```
```

Segmentation fault at address 0x1003ef0
/home/luna/.zvm/0.13.0/lib/compiler_rt/memset.zig:19:14: 0x10e6f50 in memset (compiler_rt)
            d[0] = c;
             ^
/home/luna/.zvm/0.13.0/lib/std/mem/Allocator.zig:313:26: 0x103780f in free__anon_3488 (test65)
    @memset(non_const_ptr[0..bytes_len], undefined);
                         ^
/home/luna/test-zig/test65.zig:8:19: 0x10376d6 in main (test65)
    allocator.free(coolData);
                  ^
/home/luna/.zvm/0.13.0/lib/std/start.zig:524:37: 0x1037654 in posixCallMainAndExit (test65)
            const result = root.main() catch |err| {
                                    ^
/home/luna/.zvm/0.13.0/lib/std/start.zig:266:5: 0x1037191 in _start (test65)
    asm volatile (switch (native_arch) {
    ^
???:?:?: 0x0 in ??? (???)
fish: Job 1, 'zig run test65.zig' terminated by signal SIGABRT (Abort)
```